### PR TITLE
Add TFGPT2ForSequenceClassification based on DialogRPT

### LINKED ISSUE #7622: Implement a TF2 version of `GPT2ForSequenceClassification`
# 🚀 Feature request

The architecture `GPT2ForSequenceClassification` was added in #7501 in PyTorch. It would be great to have it in TensorFlow as well.

Below is a list of items to follow to make sure the integration is complete:

- Implement `TFGPT2ForSequenceClassification` in `modeling_tf_gpt2.py`
- Test that architecture in `tests/test_modeling_tf_gpt2.py`
- Add that architecture to `__init__.py` and ` docs/source/model_doc/gpt2.rst`.

Taking a look at the code changes in #7501 would be a good start, as this PR would essentially be a TF2 copy of it.

A very good first issue to get acquainted with the library and its architectures!

### PR DESCRIPTION
**Pull request by spatil6**
# What does this PR do?

This PR implements TFGPT2ForSequenceClassification in order to support DialogRPT.
Strongly based on modifications made in #7501
Fixes #7622 

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

@LysandreJik Please review this PR, let me know if there is anything that should be changed =)

### CODE REVIEW

File: src/transformers/models/gpt2/modeling_tf_gpt2.py [hunk 1]
```diff
@@ -811,3 +816,164 @@ def call(
...
+
+    Since it does classification on the last token, it requires to know the position of the last token. If a
+    :obj:`pad_token_id` is defined in the configuration, it finds the last token that is not a padding token in each
+    row. If no :obj:`pad_token_id` is defined, it simply takes the last value in each row of the batch. Since it cannot
+    guess the padding tokens when :obj:`inputs_embeds` are passed instead of :obj:`input_ids`, it does the same (take
+    the last value in each row of the batch).
+    """,
+    GPT2_START_DOCSTRING,
+)
+class TFGPT2ForSequenceClassification(TFGPT2PreTrainedModel, TFSequenceClassificationLoss):
```
Comment (jplu) [inline src/transformers/models/gpt2/modeling_tf_gpt2.py:836]: Be careful, `TFSequenceClassificationLoss` takes only into account `-100` as pad token id. So either you assume everywhere to have this same value, either you should redefine yourself the loss computation.

File: src/transformers/models/gpt2/modeling_tf_gpt2.py [hunk 2]
```diff
@@ -811,3 +816,164 @@ def call(
...
+                    f"{self.__class__.__name__} will not detect padding tokens in `inputs_embeds`. Results may be "
+                    f"unexpected if using padding tokens in conjuction with `inputs_embeds.`"
+                )
+        loss = None
+        p_logits = None
+        if labels is not None:
+            if input_ids is not None:
+                batch_size, sequence_length = input_ids.shape[:2]
+            else:
+                batch_size, sequence_length = inputs_embeds.shape[:2]
```
Comment (jplu) [inline src/transformers/models/gpt2/modeling_tf_gpt2.py:947]: Use the `shape_list` function instead of the `.shape`.

File: src/transformers/models/gpt2/modeling_tf_gpt2.py [hunk 3]
```diff
@@ -811,3 +816,164 @@ def call(
...
+
+                def get_seq_element(sequence_position, input_batch):
+                    return tf.strided_slice(
+                        input_batch, [sequence_position - 1, 0], [sequence_position, input_batch.shape[-1]], [1, 1]
+                    )
+
+                result = tf.map_fn(
+                    fn=lambda t: get_seq_element(t[0], t[1]), elems=[sequence_lengths, logits], dtype="float"
+                )
+                p_logits = tf.reshape(result, [result.shape[0], result.shape[-1]])
```
Comment (jplu) [inline src/transformers/models/gpt2/modeling_tf_gpt2.py:962]: Same here.

File: src/transformers/models/gpt2/modeling_tf_gpt2.py [hunk 4]
```diff
@@ -811,3 +816,164 @@ def call(
...
+        attention_mask=None,
+        token_type_ids=None,
+        position_ids=None,
+        head_mask=None,
+        inputs_embeds=None,
+        labels=None,
+        use_cache=None,
+        output_attentions=None,
+        output_hidden_states=None,
+        return_dict=None,
```
Comment (jplu) [inline src/transformers/models/gpt2/modeling_tf_gpt2.py:868]: The list is not in the proper order, same think for the input order processing below. Look at the `TFGPT2LMHeadModel` class to see an example.

**Review by jplu — COMMENTED** [review]
Thank you very much for this very nice addition!!

I left few comments on it. Also can you run the following piece of code and tell me if it works properly:
```
import tensorflow as tf
from transformers import GPT2Tokenizer, TFGPT2ForSequenceClassification

model = tf.function(TFGPT2ForSequenceClassification.from_pretrained("microsoft/dialogrpt"))
tokenizer = GPT2Tokenizer.from_pretrained("microsoft/dialogrpt")
inputs = tokenizer("Hello", return_tensors="tf")
model(inputs)
```

@LysandreJik I would recommend as well to wait a bit that the new input processing to be merged.

**Comment by spatil6** [comment]
Thanks for review @jplu . I'll update my code with review comments and new input processing.

File: src/transformers/models/gpt2/modeling_tf_gpt2.py (hunk 3 above)
Comment (spatil6) [reply src/transformers/models/gpt2/modeling_tf_gpt2.py]: Done

File: src/transformers/models/gpt2/modeling_tf_gpt2.py (hunk 2 above)
Comment (spatil6) [reply src/transformers/models/gpt2/modeling_tf_gpt2.py]: Done

File: src/transformers/models/gpt2/modeling_tf_gpt2.py (hunk 4 above)
Comment (spatil6) [reply src/transformers/models/gpt2/modeling_tf_gpt2.py]: Done

**Comment by spatil6** [comment]
> Thank you very much for this very nice addition!!
> 
> I left few comments on it. Also can you run the following piece of code and tell me if it works properly:
> 
> ```
> import tensorflow as tf
> from transformers import GPT2Tokenizer, TFGPT2ForSequenceClassification
> 
> model = tf.function(TFGPT2ForSequenceClassification.from_pretrained("microsoft/dialogrpt"))
> tokenizer = GPT2Tokenizer.from_pretrained("microsoft/dialogrpt")
> inputs = tokenizer("Hello", return_tensors="tf")
> model(inputs)
> ```
> 
> @LysandreJik I would recommend as well to wait a bit that the new input processing to be merged.

output

**Comment by spatil6** [comment]
Hello @jplu and @LysandreJik ,
I have refactored code as per review comments and added new input processing as well.

Kindly review.

File: src/transformers/models/gpt2/modeling_tf_gpt2.py (hunk 1 above)
Comment (jplu) [reply src/transformers/models/gpt2/modeling_tf_gpt2.py:882]: Apparently you forgot this comment :)

**Review by jplu — COMMENTED** [review]
Much better!! Thanks for the updates.

There is still one comment to be addressed and the tests to fix.

File: src/transformers/models/gpt2/modeling_tf_gpt2.py (hunk 1 above)
Comment (spatil6) [reply src/transformers/models/gpt2/modeling_tf_gpt2.py:882]: TFGPT2ForSequenceClassification uses the last token in order to do the classification. So for given task, if a pad_token_id is defined in the configuration, it finds the last token that is not a pad token in each row, If no pad_token_id is defined, it simply takes the last value in each row of the batch.

So I think TFSequenceClassificationLoss loss function should work, as it takes single last token which is not pad token. 

Let me know your views on it. 

@LysandreJik has already defined pad_token_id: 50256 for this model.
https://github.com/huggingface/transformers/issues/7493#issuecomment-706066964

Comment (spatil6) [reply src/transformers/models/gpt2/modeling_tf_gpt2.py:882]: image

This is snapshot of TFSequenceClassificationLoss, Not any mention of pad token id : -100 is aregetting ignored. 

There are other loss function like TFTokenClassificationLoss,TFNextSentencePredictionLoss where toke id : -100 is getting ignored.

Comment (jplu) [reply src/transformers/models/gpt2/modeling_tf_gpt2.py:882]: Ok, perfect for me then!

**Comment by spatil6** [comment]
> Much better!! Thanks for the updates.
> 
> There is still one comment to be addressed and the tests to fix.

@jplu tests are also fixed now.

File: docs/source/model_doc/gpt2.rst [hunk 5]
```diff
@@ -114,3 +114,9 @@ TFGPT2DoubleHeadsModel
 
 .. autoclass:: transformers.TFGPT2DoubleHeadsModel
     :members: call
+
+TFGPT2ForSequenceClassification
```
Comment (LysandreJik) [inline docs/source/model_doc/gpt2.rst:118]: Could you also add the `TFSequenceClassifierOutputWithPast` here in the model-specific outputs?

**Review by LysandreJik — APPROVED** [review]
LGTM! Tested it out locally, it works great :) Thanks @spatil6!

**Comment by jplu** [comment]
@spatil6 we have merged today a PR that updates the way the booleans are processed. You can see an example in the TF BERT file for example, can you rebase and proceed to the same changes please. It would be awesome if you could do it!

File: docs/source/model_doc/gpt2.rst (hunk 5 above)
Comment (spatil6) [reply docs/source/model_doc/gpt2.rst:118]: Ok I'll do it.

**Comment by spatil6** [comment]
> @spatil6 we have merged today a PR that updates the way the booleans are processed. You can see an example in the TF BERT file for example, can you rebase and proceed to the same changes please. It would be awesome if you could do it!

Sure, will do that.

File: docs/source/model_doc/gpt2.rst (hunk 5 above)
Comment (spatil6) [reply docs/source/model_doc/gpt2.rst:118]: Done.

**Comment by spatil6** [comment]
> @spatil6 we have merged today a PR that updates the way the booleans are processed. You can see an example in the TF BERT file for example, can you rebase and proceed to the same changes please. It would be awesome if you could do it!

Changes are done.

But ci is failing for T5 model, what need to be done here, please advice @jplu @LysandreJik

**Review by LysandreJik — APPROVED** [review]
The new inputs processing look good to me. If it looks good to you @jplu, feel free to merge!

**Review by jplu — APPROVED** [review]
LGTM!! Good work!!
